janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/AdjacentVertexOptimizerStrategy.java

```diff
@@ -0,0 +1,160 @@
+import org.apache.tinkerpop.gremlin.process.traversal.Traversal;
+import org.apache.tinkerpop.gremlin.process.traversal.TraversalStrategy;
+import org.apache.tinkerpop.gremlin.process.traversal.step.filter.FilterStep;
+import org.apache.tinkerpop.gremlin.process.traversal.step.filter.HasStep;
+import org.apache.tinkerpop.gremlin.process.traversal.step.map.EdgeOtherVertexStep;
+import org.apache.tinkerpop.gremlin.process.traversal.step.map.EdgeVertexStep;
+import org.apache.tinkerpop.gremlin.process.traversal.step.map.VertexStep;
+import org.apache.tinkerpop.gremlin.process.traversal.step.util.HasContainer;
+import org.apache.tinkerpop.gremlin.process.traversal.strategy.AbstractTraversalStrategy;
+import org.apache.tinkerpop.gremlin.process.traversal.util.TraversalHelper;
+import org.apache.tinkerpop.gremlin.structure.Direction;
+import org.apache.tinkerpop.gremlin.structure.Edge;
+import org.apache.tinkerpop.gremlin.structure.Vertex;
+
+/**
+ * @author Florian Grieskamp (Florian.Grieskamp@gdata.de)
+ */
+public abstract class AdjacentVertexOptimizerStrategy<StepType extends FilterStep<?>>
+    extends AbstractTraversalStrategy<TraversalStrategy.ProviderOptimizationStrategy>
+    implements TraversalStrategy.ProviderOptimizationStrategy {
+
+    StepType stepType;
+
+    protected enum OptimizablePosition {
+        NONE,
+        V2V_ID,    // vertex-to-vertex step with following filter on id
+        V2E_E2V_ID // vertex-to-edge step with following edge-to-vertex-step and filter on id
+    }
+
+    protected void optimizeStep(StepType step) {
+        if (isOptimizableByPredicate(step)) {
+            OptimizablePosition pos = isOptimizableByPosition(step);
+            P<?> predicate = parsePredicate(step);
+            replaceSequence(step, pos, predicate);
+        }
+    }
+
+    private boolean isOptimizableByPredicate(StepType step) {
+        return isValidPredicate(parsePredicate(step));
+    }
+
+    protected abstract P<?> parsePredicate(StepType step);
+
+    protected abstract boolean isValidPredicate(P<?> predicate);
+
+    private OptimizablePosition isOptimizableByPosition(StepType originalStep) {
+        Step<?, ?> predecessor = originalStep.getPreviousStep();
+        Step<?, ?> prePredecessor = predecessor.getPreviousStep();
+
+        // match predecessing out(), in() or both() steps
+        if (predecessor instanceof VertexStep<?>) {
+            if (((VertexStep<?>) predecessor).returnsVertex()) {
+                return OptimizablePosition.V2V_ID;
+            }
+            return OptimizablePosition.NONE;
+        }
+
+        // match predecessing inV(), outV() or otherV() steps
+        if (predecessor instanceof EdgeVertexStep || predecessor instanceof EdgeOtherVertexStep) {
+            if (prePredecessor instanceof VertexStep) {
+                // predecessor hast to operate on an edge type
+                if (((VertexStep<?>) prePredecessor).returnsEdge()) {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [These nested if statements could be combined](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=5005131)